pysense/lib/LIS2HH12.py

```diff
@@ -95,34 +92,59 @@ def pitch(self):
         rad = -math.atan2(y, (math.sqrt(x*x + z*z)))
         return (180 / math.pi) * rad
 
+    def set_register(self, register, value, offset, mask):
+        reg = bytearray(self.i2c.readfrom_mem(ACC_I2CADDR, register, 1))
+        reg[0] &= ~(mask << offset)
+        reg[0] |= ((value & mask) << offset)
+        self.i2c.writeto_mem(ACC_I2CADDR, register, reg)
+
     def set_full_scale(self, scale):
-        self.i2c.readfrom_mem_into(ACC_I2CADDR , CTRL4_REG, self.reg)
-        self.reg[0] &= ~0b00110000
-        self.reg[0] |= (scale & 3) << 4
-        self.i2c.writeto_mem(ACC_I2CADDR , CTRL4_REG, self.reg)
+        self.set_register(CTRL4_REG, scale, 4, 3)
         self.full_scale = scale
 
     def set_odr(self, odr):
-        self.i2c.readfrom_mem_into(ACC_I2CADDR , CTRL1_REG, self.reg)
-        self.reg[0] &= ~0b01110000
-        self.reg[0] |= (odr & 7) << 4
-        self.i2c.writeto_mem(ACC_I2CADDR , CTRL1_REG, self.reg)
+        self.set_register(CTRL1_REG, odr, 4, 7)
         self.odr = odr
 
+    def set_high_pass(self, hp):
+        self.set_register(CTRL2_REG, 1 if hp else 0, 2, 1)
+
     def enable_activity_interrupt(self, threshold, duration, handler=None):
         # Threshold is in mg, duration is ms
         self.act_dur = duration
 
-        _ths = int((threshold * self.scales[self.full_scale]) / 2000 / 128) & 0x7F
+        if threshold > self.scales[self.full_scale]:
+            error = "threshold %d exceeds full scale %d" % (thresold, self.scales[self.full_scale])
+            print(error)
+            raise ValueError(error)
+
+        if threshold < self.scales[self.full_scale] / 128:
+            error = "threshold %d below resolution %d" % (thresold, self.scales[self.full_scale]/128)
+            print(error)
+            raise ValueError(error)
+
+        if duration > 255 * 1000 * 8 / self.odrs[self.odr]:
+            error = "duration %d exceeds max possible value %d" % (duration, 255 * 1000 * 8 / self.odrs[self.odr])
+            print(error)
+            raise ValueError(error)
+
+        if duration < 1000 * 8 / self.odrs[self.odr]:
+            error = "duration %d below resolution %d" % (duration, 1000 * 8 / self.odrs[self.odr])
+            print(error)
+            raise ValueError(error)
+
+        _ths = int(127 * threshold / self.scales[self.full_scale]) & 0x7F
```

Comment:
The datasheet states the LSB is equal to full scale/128 so surely this should `128` not `127`

Reply:
The data sheet has an incorrect title for that column. The application note available here http://www.st.com/content/ccc/resource/technical/document/application_note/60/52/bd/69/28/f4/48/2b/DM00165265.pdf/files/DM00165265.pdf/jcr:content/translations/en.DM00165265.pdf correctly states full scale/128 is the resolution of this field. I originally used 128, but then if you enter the full scale value you get 128, when you should clearly have 127 (the max value for that field). It shouldn't matter much though, the noise on those values is probably higher than the difference we could get with 127 vs 128.